lib/src/component_declaration/flux_component.dart

```diff
@@ -76,46 +101,101 @@ abstract class FluxUiComponent<TProps extends FluxUiProps> extends UiComponent<T
+  // Ignore this warning to work around https://github.com/dart-lang/sdk/issues/29860
+  // ignore: must_call_super
+  void componentWillMount();
+
+  @mustCallSuper
+  @override
+  // Ignore this warning to work around https://github.com/dart-lang/sdk/issues/29860
+  // ignore: must_call_super
+  void componentWillReceiveProps(Map prevProps);
+
+  @mustCallSuper
+  @override
+  void componentDidUpdate(Map prevProps, Map prevState);
+
+  @mustCallSuper
+  @override
+  void componentWillUnmount();
+}
 
 /// Helper mixin to keep [FluxUiComponent] and [FluxUiStatefulComponent] clean/DRY.
 ///
 /// Private so it will only get used in this file, since having lifecycle methods in a mixin is risky.
-abstract class _FluxComponentMixin<TProps extends FluxUiProps> implements BatchedRedraws {
-  TProps get props;
-
+abstract class _FluxComponentMixin<TProps extends FluxUiProps> implements BatchedRedraws, UiComponent<TProps> {
   /// List of store subscriptions created when the component mounts.
   ///
   /// These subscriptions are canceled when the component is unmounted.
-  List<StreamSubscription> _subscriptions = [];
+  List<StreamSubscription> _subscriptions;
+
+  bool get _areStoreHandlersBound => _subscriptions != null;
+
+  /// Subscribe to all applicable stores.
+  ///
+  /// [Store]s returned by [redrawOn] will have their triggers mapped directly to this component's
+  /// redraw function.
+  ///
+  /// [Store]s included in the [getStoreHandlers] result will be listened to and wired up to their
+  /// respective handlers.
+  void _bindStoreHandlers() {
+    if (_areStoreHandlersBound) {
+      throw new StateError('Store handlers are already bound');
+    }
 
-  void componentWillMount() {
-    /// Subscribe to all applicable stores.
-    ///
-    /// [Store]s returned by [redrawOn] will have their triggers mapped directly to this components
-    /// redraw function.
-    ///
-    /// [Store]s included in the [getStoreHandlers] result will be listened to and wired up to their
-    /// respective handlers.
     Map<Store, StoreHandler> handlers = new Map.fromIterable(redrawOn(),
         value: (_) => (_) => redraw())..addAll(getStoreHandlers());
 
+    _subscriptions = <StreamSubscription>[];
     handlers.forEach((store, handler) {
       StreamSubscription subscription = store.listen(handler);
       _subscriptions.add(subscription);
     });
   }
 
+  /// Cancel all store subscriptions.
+  void _unbindStoreHandlers() {
+    if (!_areStoreHandlersBound) return;
+
+    for (var subscription in _subscriptions) {
+      subscription?.cancel();
+    }
+
+    _subscriptions = null;
+  }
+
+  @override
+  void componentWillMount() {
+    _bindStoreHandlers();
+  }
+
+  @override
+  void componentWillReceiveProps(Map prevProps) {
+    // Unbind store handlers so they can be re-bound in componentDidUpdate
+    // once the new props are available, ensuring the values returned by [redrawOn]
+    // are not outdated.
+    _unbindStoreHandlers();
```

Comment:
We'd have to compare the equality of the return value of `redrawOn` and `getStoreHandlers`, and we don't have access to those updated values until `componentDidUpdate` (since those functions often rely on values within `props`). So, we'd have to store the old values in here and then check them and null them out in `componentDidUpdate`.

We can do that, but I figured that unbinding/rebinding was cheap enough that it was worth doing so to avoid that kind of comparison.